### ISSUE TITLE
multi-material implementation

### ISSUE THREAD

**Issue by cjoycd** [CEED/Laghos#47]
I am trying to implement multi-material hydrodynamics.
        I have few doubts and problems regarding its implementation in Laghos.

        1) Suppose ParGridFunction say "rhom" & VectorFunctionCoefficient say "rhom_coeff"
           are set for material densities and similarly for ParGridFunction "eta" &
           VectorFunctionCoefficient "eta_coeff" for material volume fractions.
           Is there any module/class/function exists in MFEM which can be used for derived
           ParGridFunction & GridFunctionCoefficient; say "rho" & "rho_coeff"
           in which rho = sum_m (eta * rhom) ? Basically I want to know, how a derived
           ParGridFunction & GridFunctionCoefficient which can be sum and/or product
           of already defined ParGridFunction & GridFunctionCoefficient obtained ?

        2) For multi-material formulation what could be a best or efficient way for
           computing force matrix, F_k (for each material k) and F = sum_k F_k
           in the AssembleForceMatrix() --> ForceIntegrator for the full assembly case.
           Is it necessary to call the AssembleForceMatrix() for F = sum_k F_k for velocity
           update and several such calls (for each k) for F_k which will be used for
           material energy (ek) update. I want an efficient way in which the
           AssembleForceMatrix() is called only once and both F = sum_k F_k and F_k (for each k)
           are computed and storred. Can anyone please suggest.

        3) Also, the energy mass matrices should also be a function of material index "k".
           How the energy mass matrix Me(i) and its inverse Me_inv(i) be made as a function
           of material index "k". Is it effient to define the index "i" --> "nzones * nmc()",
           where nmc is the total number of materials in a zone.

        4) Is there any simple template already available for full assembly case to start with
           or is there any plan to extend Laghos with this capability in near future ?

**Comment by vladotomov** [comment]
Hi Sijoy,

1 - There probably isn't anything that you can directly, as these sums / products are very specific for the application. However you can always define new `Coefficient` classes that depend on other `Coefficients`, or define custom projections for the `GridFunctions`.

2 - The fastest way to get something working is to define a `MixedBilinearForm` for each material, which has the material's specific `ForceOperator`, and another `MixedBilinearForm` that has all material `ForceOperators` for the sum. As you say, this will do repeated calculations (but it will work). For something more efficient, you'll probably need to define something that goes over a set of `MixedBilinearForms` and sums their element matrices to form another `MixedBilinearForm` for the velocity equation.

3 - Yes, you will need material-specific energy mass matrices, so these must be stored in some structure. I don't think you can avoid the extra memory, unless you recompute every time the same thing.

4 - Currently we're not planning to extend Laghos to multiple materials and mixed zones. Sorry, there is no simple template.

Let us know how this goes, it would be great to see Laghos extended to multiple materials by an external user.

Thanks,
Vladimir

**Comment by cjoycd** [comment]
Dear Vladimir,
Thank you for your prompt comments, I will go ahead experimenting. 

Regards,
--Sijoy 






> Hi Sijoy,
> 
> 1 - There probably isn't anything that you can directly, as these sums / products are very specific for the application. However you can always define new `Coefficient` classes that depend on other `Coefficients`, or define custom projections for the `GridFunctions`.
> 
> 2 - The fastest way to get something working is to define a `MixedBilinearForm` for each material, which has the material's specific `ForceOperator`, and another `MixedBilinearForm` that has all material `ForceOperators` for the sum. As you say, this will do repeated calculations (but it will work). For something more efficient, you'll probably need to define something that goes over a set of `MixedBilinearForms` and sums their element matrices to form another `MixedBilinearForm` for the velocity equation.
> 
> 3 - Yes, you will need material-specific energy mass matrices, so these must be stored in some structure. I don't think you can avoid the extra memory, unless you recompute every time the same thing.
> 
> 4 - Currently we're not planning to extend Laghos to multiple materials and mixed zones. Sorry, there is no simple template.
> 
> Let us know how this goes, it would be great to see Laghos extended to multiple materials by an external user.
> 
> Thanks,
> Vladimir

**Comment by cjoycd** [comment]
Dear Vladimir,

Finally, I could implement the multi-material formulation. At present It is implemented with constant volume fraction and to full assembly (FA). Its seems to be working and done with all basic conservation checks. I would like to implement now the volume fraction update (using multi-material closure model described by Dobrev et al., Int. J. Num. Meth. Fluids, 82 (2016), pp. 689–706). I have a doubt here. The volume fractions are updated at each quadratic point. At t=0, they are interpolated from the the GridFunction. Later the volume fractions are updated at quadrature level. Is it sufficient to have a simple explicit update of volume fractions inside the UpdateQuadratureData module or is it necessary to integrate using higher order schemes as it is done for state vectors (S --> x,v,e). In both cases, I think the volume fraction update must take care, repetition of failed steps leading to additional storage of old volume fractions at quadrature points for a fresh calculation.  Another issue is to plot the total mass specific energy [e = sum_k (rho_k  * eta_k * e_k ) / rho]. Probably, I have to assign new energy ParGridFunction from existing quadrature values eta_k, rho_k & r_k and rho = sum_k (eta_k, rho_k). Is this right ? 
 
Thanks for your help. 

Regards
Sijoy C.D.

**Comment by vladotomov** [comment]
Hi Sijoy,

In the IJNMF paper, volume fractions (aka material indicators) are integrated in time using the same time integrator as the state vector, i.e., the indicator values become part of the state. Yes, repetition of failed steps can be handled the same way.

Yes, to visualize total mass specific internal energy `e` you must project to some new `ParGridFunction` using the existing quadrature point values; sounds right.

Great progress, send us some results!

Thanks,
Vladimir

**Comment by cjoycd** [comment]
Dear Vladimir,

I am facing a problem initializing the mixed zones in a computational domain. I have initialized basically three Vector GridFunctions (dim = no. of materials); one for volume fraction, other for density of each material and finally for energy of each material.  These are initialized using VectorFunctionCoefficient (with functions supplemented say eta0, rho0 & e0). I have noticed the coefficients are sampled in the function supplied is for x,y,z locations at thermodynamics dofs. This is fine. However, while getting values at the quadrature points they are interpolated from these GridFunctions. For volume fractions it looks fine as one get a smooth variation of volume fraction max (1.0) to min (0.0) within the cell for each material at the quadrature points. However, at quadrature points the similar variation occurs for material densities and energies (which can be zero at times). What is required at a quadrature point is a non-zero or constant zone value for material density and energy whenever volume fraction of that material is non-zero.  I had read in your IJNMF paper as some extended values to be managed at these locations wherever volume fraction for a material is > 0.0. If that is the case how it is exactly done ? Do we need to take care this at first step only ? For density as it is defined at the quadrature point at t=0 from a gridfunction and evolved at these points following strong mass conservation. How this can be done conservatively for energies ? 

Thank you for your time in advance

Sijoy

**Comment by vladotomov** [comment]
Hello Sijoy,

Yes, in our IJNMF paper we use extended values for density and energy in the mixed zones. As written in section 4.1, this is done only at initial time. It does not affect the method at later times, hence the mass and energy are still conserved for each material in the evolution, compared to the initial values.

Thanks,
Vladimir

**Comment by cjoycd** [comment]
Thanks Vladimir

**Comment by cjoycd** [comment]
Dear Vladimir,

I could finally implement it..although the initialization is tricky.. It doesn't matter in most of the simulations as the mixed cell situations will arise after a remap-step. I just wanted a multi-material formulation is implemented & tested for a remap stage which hopefully will be implemented in future :).   
  
Thank you for your support
--Sijoy C.D.

**Comment by cjoycd** [comment]
> Hi Sijoy,
> 
> In the IJNMF paper, volume fractions (aka material indicators) are integrated in time using the same time integrator as the state vector, i.e., the indicator values become part of the state. Yes, repetition of failed steps can be handled the same way.
> 
> Yes, to visualize total mass specific internal energy `e` you must project to some new `ParGridFunction` using the existing quadrature point values; sounds right.
> 
> Great progress, send us some results!
> 
> Thanks, Vladimir

Dear Vladimir,

Finally, an axis-symmetric multi-material implementation is completed following IJNMF ALE paper. I have used FCT based local L2 projection prior and after the remap, as described in the paper "HIGH-ORDER MULTI-MATERIAL ALE HYDRODYNAMICS, SIAM Journal on Scientific Computing (2016). For remap stage, product synchronized FCT algorithm is used. Although the multi-material test results are satisfactory, there are energy and mass conservation errors (This can be due to some minor implementation errors and bugs -- which will be investigated). For example, for three-material tripple point problem, a mass conservation error of 0.01-0.05 % and an energy conservation error of ~1 % are seen in the simulation. These errors are, however, tends to reduce with mesh refinement. The error introduced in the total energy is mainly due the errors in the K.E. ( a velocity/momentum field remap without any monotonicity treatment). Are these conservation errors expected/acceptable for the typical implementation/algorithm described in the ALE paper? Do I need to use any monotonicity treatment (in continuous FE space) for velocity components as well?  I have attached a typical density plot for a three material tripple point problem for which the energy and mass errors are mentioned.

Regards
--Sijoy C.D.
